File: app/src/main/java/com/fusoft/walkboner/database/funcions/userProfile/GetUserLikedPosts.java
package com.fusoft.walkboner.database.funcions.userProfile;

import android.content.Context;
import android.util.Log;
import android.widget.Toast;

import androidx.annotation.NonNull;

import com.fusoft.walkboner.models.Post;
import com.google.android.gms.tasks.OnFailureListener;
import com.google.android.gms.tasks.OnSuccessListener;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseUser;
import com.google.firebase.firestore.DocumentSnapshot;
import com.google.firebase.firestore.FirebaseFirestore;
import com.google.firebase.firestore.Query;
import com.google.firebase.firestore.QuerySnapshot;

import java.util.ArrayList;
import java.util.List;

public class GetUserLikedPosts {

    FirebaseFirestore firestore = FirebaseFirestore.getInstance();
    FirebaseAuth auth = FirebaseAuth.getInstance();
    FirebaseUser user = auth.getCurrentUser();

    List<Post> posts = new ArrayList<>();

    int a = 0;

    private void log(String message) {
        Log.e("UPostsLog", "______________");
        Log.e("UPostsLog", message);
    }

    public void get(Context context, UserLikedPostsListener listener) {
        log("Getting User Liked Posts From Database...");
        firestore.collection("users").whereEqualTo("userUid", user.getUid()).get().addOnSuccessListener(new OnSuccessListener<QuerySnapshot>() {
            @Override
            public void onSuccess(QuerySnapshot queryDocumentSnapshots) {
                log("User Finded!");
                DocumentSnapshot userDocument = queryDocumentSnapshots.getDocuments().get(0);

                userDocument.getReference().collection("likedPosts").get().addOnSuccessListener(new OnSuccessListener<QuerySnapshot>() {
                    @Override
                    public void onSuccess(QuerySnapshot queryDocumentSnapshots) {
                        log("Finded Liked Posts with size " + queryDocumentSnapshots.getDocuments().size());
                        for (DocumentSnapshot likedPostsUserDocument : queryDocumentSnapshots.getDocuments()) {
                            log("Getting post data for post with UID " + likedPostsUserDocument.getString("postUid"));
                            firestore.collection("posts").document(likedPostsUserDocument.getString("postUid")).get().addOnSuccessListener(new OnSuccessListener<DocumentSnapshot>() {
                                @Override
                                public void onSuccess(DocumentSnapshot documentSnapshot) {
                                    log("Finded Post with UID " + documentSnapshot.getString("postUid"));
                                    Post post = new Post();
                                    post.setUserUid(documentSnapshot.getString("userUid"));
                                    post.setPostUid(documentSnapshot.getString("postUid"));
                                    post.setPostDocumentUid(documentSnapshot.getId());
                                    post.setPostDescription(documentSnapshot.getString("postDescription"));
                                    post.setPostImage(documentSnapshot.getString("postImage"));
                                    post.setCreatedAt(documentSnapshot.getString("createdAt"));
                                    post.setShowsCelebrity(documentSnapshot.getString("showsCelebrity"));
                                    post.setAllowComments(Boolean.parseBoolean(documentSnapshot.getString("allowComments")));
                                    post.setUserLikedPost(true);

                                    List<String> likesList = new ArrayList<>();
                                    firestore.collection("posts").document(documentSnapshot.getId()).collection("likes").get().addOnSuccessListener(queryDocumentSnapshots1 -> {
                                        for (DocumentSnapshot docLikes : queryDocumentSnapshots1.getDocuments()) {
                                            likesList.add(docLikes.getString("userUid"));
                                            if (docLikes.getString("userUid").contentEquals(user.getUid())) {
                                                post.setUserLikedPost(true);
                                            }
                                        }
                                    });
                                    post.setPostLikes(likesList);
                                    log("Post with description | " + post.getPostDescription() + " | finded!");
                                    posts.add(post);
                                    listener.OnLoaded(posts);
                                }
                            }).addOnFailureListener(new OnFailureListener() {
                                @Override
                                public void onFailure(@NonNull Exception e) {
                                    log("Błąd 82: " + e.getMessage());
                                    listener.OnError("Nie udało się uzyskać szczegółów posta: " + e.getMessage());
                                }
                            });
                        }
                        log("listener invoked");
                    }
                }).addOnFailureListener(new OnFailureListener() {
                    @Override
                    public void onFailure(@NonNull Exception e) {
                        listener.OnError("Nie udało się przechwycić polubionych postów użytkownika:\n" + e.getMessage());
                    }
                });
            }
        }).addOnFailureListener(new OnFailureListener() {
            @Override
            public void onFailure(@NonNull Exception e) {
                listener.OnError("Nie znaleziono użytkownika:\n" + e.getMessage());
            }
        });
    }

    public interface UserLikedPostsListener {
        void OnLoaded(List<Post> posts);

        void OnError(String error);
    }
}

File: app/src/main/java/com/fusoft/walkboner/database/funcions/userProfile/GetUserLikedPosts.kt
package com.fusoft.walkboner.database.funcions.userProfile

import android.util.Log
import com.fusoft.walkboner.models.Post
import com.fusoft.walkboner.singletons.UserSingletone
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.firestore.DocumentSnapshot
import com.google.firebase.firestore.FirebaseFirestore
import com.google.firebase.firestore.Query
import com.google.firebase.firestore.QuerySnapshot
import kotlinx.coroutines.tasks.await
import java.lang.Boolean
import kotlin.String
import kotlin.let

class GetUserLikedPosts {
    private fun log(message: String) {
        Log.e("UPostsLog", "______________")
        Log.e("UPostsLog", message)
    }

    suspend fun get(listener: UserLikedPostsListener) {
        val firestore = FirebaseFirestore.getInstance()
        val auth = FirebaseAuth.getInstance()
        val user = auth.currentUser
        val posts: MutableList<Post> = ArrayList()
        var userDocument: DocumentSnapshot? = null
        var likedPostsQuerySnapshot: QuerySnapshot? = null
        val userDocId = UserSingletone.getInstance().user.userDocumentId
        posts.clear()

        log("Getting User Liked Posts From Database...")
        firestore.collection("users").document(userDocId).get()
            .addOnSuccessListener { userDoc ->
                log("User Finded!")
                userDocument = userDoc
            }
            .addOnFailureListener {
                listener.OnError("""Nie znaleziono użytkownika:${it.message}""".trimIndent())
            }.await()

        // Check if user is null
        if (userDocument == null) {
            listener.OnError("Nie znaleziono użytkownika!")
            return
        }

        userDocument?.reference?.collection("likedPosts")
            ?.orderBy("likeAt", Query.Direction.DESCENDING)?.get()
            ?.addOnSuccessListener { queryDocumentSnapshots ->
                log("Finded Liked Posts with size " + queryDocumentSnapshots.documents.size)
                likedPostsQuerySnapshot = queryDocumentSnapshots
            }?.await()

        // Check if no liked posts
        if (likedPostsQuerySnapshot == null || likedPostsQuerySnapshot?.isEmpty == true) {
            listener.OnLoaded(null)
            return
        }

        for (likedPostsUserDocument in likedPostsQuerySnapshot?.documents!!) {
            log(
                "Getting post data for post with UID " + likedPostsUserDocument.getString(
                    "postUid"
                )
            )
            var post: Post? = null
            var postSnapshot: DocumentSnapshot? = null
            firestore.collection("posts")
                .document(likedPostsUserDocument.getString("postUid")!!).get()
                .addOnSuccessListener { documentSnapshot ->
                    log("Finded Post with UID " + documentSnapshot.getString("postUid"))
                    postSnapshot = documentSnapshot
                }
                .addOnFailureListener { e ->
                    log("Błąd 82: " + e.message)
                    listener.OnError("Nie udało się uzyskać szczegółów posta: " + e.message)
                }.await()
            post = Post()
            post.userUid = postSnapshot?.getString("userUid")
            post.postUid = postSnapshot?.getString("postUid")
            post.postDocumentUid = postSnapshot?.id
            post.postDescription =
                postSnapshot?.getString("postDescription")
            post.postImage = postSnapshot?.getString("postImage")
            post.createdAt = postSnapshot?.getString("createdAt")
            post.showsCelebrity =
                postSnapshot?.getString("showsCelebrity")
            post.isAllowComments =
                Boolean.parseBoolean(postSnapshot?.getString("allowComments"))
            post.isUserLikedPost = true
            val likesList: MutableList<String?> = java.util.ArrayList()
            var likesQueryDocumentSnapshot: QuerySnapshot? = null
            postSnapshot?.id?.let {
                firestore.collection("posts").document(it)
                    .collection("likes").get()
                    .addOnSuccessListener { queryDocumentSnapshots1: QuerySnapshot ->
                        likesQueryDocumentSnapshot = queryDocumentSnapshots1
                    }
            }?.await()

            for (docLikes in likesQueryDocumentSnapshot?.documents!!) {
                likesList.add(docLikes.getString("userUid"))
                if (docLikes.getString("userUid")
                        .contentEquals(user!!.uid)
                ) {
                    post.isUserLikedPost = true
                }
            }

            post.postLikes = likesList
            log("Post with description | " + post.postDescription + " | finded!")
            posts.add(post)
        }

        listener.OnLoaded(posts)
        log("listener invoked")
    }

    interface UserLikedPostsListener {
        fun OnLoaded(posts: List<Post>?)
        fun OnError(error: String?)
    }
}